### PR TITLE
:tada: Destination Bigquery denormalization - 9345

### PR DESCRIPTION
**Pull request by andriikorotkov**
## What
Updated destination-bigquery-denormalize and moved all logic to the destination-bigquery.

## How
Now, if you need to update destination-bigquery-denormalize, all changes need to be added to destination-bigquery. Only the main class and tests remain in destination-bigquery-denormalize. All logic is transferred to destination-bigquery and in destination-bigquery-denormalize it is used by means of a wrapper.

## Recommended reading order
1. `x.java`

## 🚨 User Impact 🚨
To update destination-bigquery-denormalize - you need to add changes to destination-bigquery, and you also need to update version destination-bigquery.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 1]
```diff
@@ -183,17 +194,28 @@ public AirbyteMessageConsumer getConsumer(final JsonNode config,
...
+              UploaderType.AVRO, new GcsAvroBigQueryRecordFormatter(jsonSchema, getNamingResolver()));
+    }
   }
 
   protected String getTargetTableName(final String streamName) {
-    return namingResolver.getRawTableName(streamName);
+    if (denormalize) {
+    // This BigQuery destination does not write to a staging "raw" table but directly to a normalized
+    // table
+    return getNamingResolver().getIdentifier(streamName);
```
Comment (mkhokh-33) [inline airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:215]: we do we use different approaches :  getNamingResolver() and namingResolver itself?

**Review by mkhokh-33 — CHANGES_REQUESTED** [review]

**Review by mkhokh-33 — APPROVED** [review]

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java (hunk 1 above)
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java]: Yes, this is my mistake. Earlier in bigquery-denormalized we used getter. now, we can use namingResolver directions. Changed that.

**Review by edgao — COMMENTED** [review]
is this intended to fully resolve #9345 ? or is there some followup work that's planned for a separate PR? (e.g. there's still some files in the destination-bigquery-denormalized directory; handling conditional normalization, etc)

**Comment by andriikorotkov** [comment]
@edgao further work for this task can be aimed only at summarizing spec destination-bigquery-denormalized and destination-bigquery to one file and the logic of using the desired specification, but at the moment it is not possible to do so. All logic from destination-bigquery-denormalized will be moved to destination-bigquery. In destination-bigquery-denormalized there is only the main class and tests.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/wrapper/WrappedDestination.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.function.Consumer;
+
+/**
+ * Wrap to use a destination as the basis for another destination.
+ * For example, using destination-bigquery for destination-bigquery-denormalized.
+ */
+public class WrappedDestination implements Destination {
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/wrapper/WrappedDestination.java:23]: Yes, I agree with you. Probably, I had a slightly different idea, but in the process of implementation I gave it up =) Yes, thanks for the remarks, I removed the extra wrapper and changed the call logic to the 'new BigQueryDestination (denormalize)'.

File: airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java [hunk 3]
```diff
@@ -4,50 +4,42 @@
...
+    final ConnectorSpecification spec = Jsons.clone(originalSpec);
+    spec.setSupportedDestinationSyncModes(asList(APPEND, OVERWRITE));
+    spec.setSupportsNormalization(false);
+    ((ObjectNode) spec.getConnectionSpecification().get("properties")).remove("transformation_priority");
+    return spec;
   }
 
   public static void main(final String[] args) throws Exception {
     final Destination destination = new BigQueryDenormalizedDestination();
+    LOGGER.info("starting destination: {}", BigQueryDenormalizedDestination.class);
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:40]: in most main connector classes, these logs are called the way I fixed. For example, for destination-mysql and destination-postgres.

File: airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java [hunk 4]
```diff
@@ -183,17 +194,28 @@ public AirbyteMessageConsumer getConsumer(final JsonNode config,
...
+      return Map.of(UploaderType.STANDARD, new DefaultBigQueryRecordFormatter(jsonSchema, getNamingResolver()),
+              UploaderType.CSV, new GcsCsvBigQueryRecordFormatter(jsonSchema, getNamingResolver()),
+              UploaderType.AVRO, new GcsAvroBigQueryRecordFormatter(jsonSchema, getNamingResolver()));
+    }
   }
 
   protected String getTargetTableName(final String streamName) {
-    return namingResolver.getRawTableName(streamName);
+    if (denormalize) {
+    // This BigQuery destination does not write to a staging "raw" table but directly to a normalized
```
[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:213]: +

[the remark this answers is not in the document]
Comment (andriikorotkov) [reply airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDestination.java:213]: Yes of course.

**Review by edgao — APPROVED** [review]
hmmm - I don't think this PR fully resolves #9345 (i.e. users will still see both bigquery destinations), and per @sherifnada it sounds like we should actually hold off on that issue for now.

this specific change makes sense though, just to reduce the code duplication. So I think we should remove the link to #9345 and merge this in.

**Review by sherifnada — CHANGES_REQUESTED** [review]
see comment above

**Comment by andriikorotkov** [comment]
Yes of course. We can put it off, and get back to this pull request at any time.

**Comment by alexandr-shegeda** [comment]
original ticket was closed https://github.com/airbytehq/airbyte/issues/11154

**Pull request closed without merging**